### PR TITLE
Check attribute usage

### LINKED ISSUE #2809: Validate usage of inline attributes
```syntax::attr::find_inline_attr``` has a FIXME saying "validate the usage of ```#[inline]``` and ```#[inline(always)]```".

### LINKED ISSUE #22746: error: internal compiler error: range_to_inttype: found ReprPacked on an enum
http://is.gd/XqFILb

```
stack backtrace:
   1:        0x10e063be3 - sys::backtrace::write::hc8e3cee73e646c590nC
   2:        0x10e091be5 - panicking::on_panic::h00b47941f5bc8a02HOL
   3:        0x10dfbb418 - rt::unwind::begin_unwind_inner::h539538ef7f909326UvL
   4:        0x10d678dcf - rt::unwind::begin_unwind::h17340398235865146743
   5:        0x10d6796fd - diagnostic::Handler::bug::h7e78f4ac01756c63w4E
   6:        0x10aef2028 - session::Session::bug::hb92b783f7b3446ffSnt
   7:        0x10a947da8 - check::check_item::h4a1b4ae849c9e701nTn
   8:        0x10aa19b73 - check_crate::closure.35867
   9:        0x10aa14911 - check_crate::hbe2eb4aab434ad7cVjC
  10:        0x10a3768db - driver::phase_3_run_analysis_passes::h50728865d3ffa6b2gHa
  11:        0x10a35bb9d - driver::compile_input::h349b039ebe75c13eGba
  12:        0x10a434857 - run_compiler::h1e4d87c6044fbf6dZbc
  13:        0x10a431cd1 - thunk::F.Invoke<A, R>::invoke::h2059937391804860278
  14:        0x10a430930 - rt::unwind::try::try_fn::h11862807381821880905
  15:        0x10e10b7e9 - rust_try_inner
  16:        0x10e10b7d6 - rust_try
  17:        0x10a431095 - thunk::F.Invoke<A, R>::invoke::h6573516661401600357
  18:        0x10e07aac3 - sys::thread::thread_start::h9e3555425c6999b2e4G
  19:     0x7fff966f1268 - _pthread_body
  20:     0x7fff966f11e5 - _pthread_body
```

### PR DESCRIPTION
**Pull request by sanxiyn**
This is technically a [breaking-change].

Fix #2809.
Fix #22746.

### CODE REVIEW

**Comment by alexcrichton** [comment]
Starting a crater build to evaluate the impact

**Comment by alexcrichton** [comment]
Looks like there's only [one regression](https://gist.github.com/alexcrichton/04b88f26def9ca38ada6), so should be easy to send a PR for!

**Comment by nagisa** [comment]
I’d replace error message wording from “should” to “may only” or “may”.

Otherwise always :+1: for such correctness PRs.

File: src/librustc/front/check_attr.rs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+                "attribute should be applied to function");
+        }
+    }
+
+    fn check_repr(&self, attr: &ast::Attribute, target: Target) {
+        let words = attr.meta_item_list();
+        if words.is_none() {
+            return;
+        }
+        let words = words.unwrap();
```
Comment (nrc) [inline src/librustc/front/check_attr.rs]: nit:
```
let words = match attr.meta_item_list() {
    Some(words) => words,
    None => {
        return;
    }
};
```

**Comment by nrc** [comment]
r=me with the style nit fixed

**Comment by sanxiyn** [comment]
Done.

**Comment by Manishearth** [comment]
IMO we should have made this a warning for one release cycle, but I guess this is fine too.

**Pull request merged**